pkg/cloud/azure/azure.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/submariner-io/cloud-prepare/pkg/ocp"
+	"github.com/submariner-io/subctl/internal/restconfig"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Config struct {
+	Gateways        int
+	InfraID         string
+	Region          string
+	OcpMetadataFile string
+	AuthFile        string
+	GWInstanceType  string
+}
+
+func RunOn(restConfigProducer *restconfig.Producer, config *Config, status reporter.Interface,
+	function func(api.Cloud, api.GatewayDeployer, reporter.Interface) error,
+) error {
+	if config.OcpMetadataFile != "" {
+		var err error
+
+		config.InfraID, config.Region, err = ReadFromFile(config.OcpMetadataFile)
+		if err != nil {
+			return status.Error(err, "Failed to read AWS information from OCP metadata file %q", config.OcpMetadataFile)
+		}
+	}
+
+	status.Start("Retrieving Azure credentials from your Azure authorization file")
+
+	err := os.Setenv("AZURE_AUTH_LOCATION", config.AuthFile)
+	if err != nil {
+		return status.Error(err, "Error locating authorization file")
+	}
+
+	subscriptionID, err := initializeFromAuthFile(config.AuthFile)
+	if err != nil {
+		return status.Error(err, "Failed to read authorization information from Azure authorization file")
+	}
+
+	status.End()
+
+	status.Start("Initializing AWS connectivity")
+
+	// This is the most recommended of several authentication options
+	// https://github.com/Azure/go-autorest/tree/master/autorest/azure/auth#more-authentication-details
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return status.Error(err, "Error getting an authorizer for Azure")
+	}
+
+	k8sConfig, err := restConfigProducer.ForCluster()
+	if err != nil {
+		return status.Error(err, "Failed to initialize a Kubernetes config")
+	}
+
+	clientSet, err := kubernetes.NewForConfig(k8sConfig.Config)
+	if err != nil {
+		return status.Error(err, "Failed to create Kubernetes client")
+	}
+
+	k8sClientSet := k8s.NewInterface(clientSet)
+
+	restMapper, err := util.BuildRestMapper(k8sConfig.Config)
+	if err != nil {
+		return status.Error(err, "Failed to create restmapper")
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(k8sConfig.Config)
+	if err != nil {
+		return status.Error(err, "Failed to create dynamic client")
+	}
+
+	msDeployer := ocp.NewK8sMachinesetDeployer(restMapper, dynamicClient)
+
+	cloudInfo := azure.CloudInfo{
+		SubscriptionID: subscriptionID,
+		InfraID:        config.InfraID,
+		Region:         config.Region,
+		BaseGroupName:  config.InfraID + "-rg",
+		Authorizer:     authorizer,
+		K8sClient:      k8sClientSet,
+	}
+
+	azureCloud := azure.NewCloud(&cloudInfo)
+
+	status.End()
+
+	gwDeployer, err := azure.NewOcpGatewayDeployer(azureCloud, msDeployer, config.GWInstanceType)
+	if err != nil {
+		return status.Error(err, "Failed to initialize a GatewayDeployer config")
+	}
+
+	return function(azureCloud, gwDeployer, status)
+}
+
+func ReadFromFile(metadataFile string) (string, string, error) {
```

Comment:
We should use the new [ReadMetadataFile function](https://github.com/submariner-io/subctl/pull/59) once it's merged.

Reply:
Done